quinn-proto/src/tests/mod.rs

```diff
@@ -2181,6 +2208,388 @@ fn packet_splitting_not_necessary_after_higher_mtu_discovered() {
+
+    let client_stats_after_connect = pair.client_conn_mut(client_ch).stats();
+    let server_stats_after_connect = pair.server_conn_mut(server_ch).stats();
+
+    // Send a packet that won't arrive right away (it will be dropped and be re-sent later)
+    pair.mtu = 0;
+    pair.client_conn_mut(client_ch).ping();
+    pair.drive_client();
+
+    // Sanity check (ping sent, no ACK received)
+    let client_stats_after_first_ping = pair.client_conn_mut(client_ch).stats();
+    assert_eq!(
+        client_stats_after_first_ping.frame_tx.ping - client_stats_after_connect.frame_tx.ping,
+        1
+    );
+    assert_eq!(
+        client_stats_after_first_ping.frame_rx.acks - client_stats_after_connect.frame_rx.acks,
+        0
+    );
+
+    // Restore the default MTU and send another ping, which will arrive earlier than the dropped one
+    pair.mtu = default_mtu;
+    pair.client_conn_mut(client_ch).ping();
+    pair.drive_client();
+    pair.drive_server();
+    pair.drive_client();
+
+    // Client sanity check (ping sent, one ACK received)
+    let client_stats_after_second_ping = pair.client_conn_mut(client_ch).stats();
+    assert_eq!(
+        client_stats_after_second_ping.frame_tx.ping - client_stats_after_connect.frame_tx.ping,
+        2
+    );
+    assert_eq!(
+        client_stats_after_second_ping.frame_rx.acks - client_stats_after_connect.frame_rx.acks,
+        1
+    );
+
+    // Server checks (single ping received, ACK sent)
+    let server_stats_after_second_ping = pair.server_conn_mut(server_ch).stats();
+    assert_eq!(
+        server_stats_after_second_ping.frame_rx.ping - server_stats_after_connect.frame_rx.ping,
+        1
+    );
+    assert_eq!(
+        server_stats_after_second_ping.frame_tx.acks - server_stats_after_connect.frame_tx.acks,
+        1
+    );
+}
+
+#[test]
+fn single_ack_eliciting_packet_with_ce_bit_triggers_immediate_ack() {
+    let _guard = subscribe();
+    let mut pair = Pair::default();
+    let (client_ch, _) = pair.connect();
+    pair.drive();
+
+    let stats_after_connect = pair.client_conn_mut(client_ch).stats();
+
+    let start = pair.time;
+
+    pair.client_conn_mut(client_ch).ping();
+
+    pair.congestion_experienced = true;
+    pair.drive_client(); // Send ping
+    pair.congestion_experienced = false;
+
+    pair.drive_server(); // Process ping, send ACK in response to congestion
+    pair.drive_client(); // Process ACK
+
+    // Sanity check: the time hasn't advanced in the meantime)
+    assert_eq!(pair.time, start);
+
+    let stats_after_ping = pair.client_conn_mut(client_ch).stats();
+    assert_eq!(
+        stats_after_ping.frame_tx.ping - stats_after_connect.frame_tx.ping,
+        1
+    );
+    assert_eq!(
+        stats_after_ping.frame_rx.acks - stats_after_connect.frame_rx.acks,
+        1
+    );
+    assert_eq!(
+        stats_after_ping.path.congestion_events - stats_after_connect.path.congestion_events,
+        1
+    );
+}
+
+#[test]
+fn ack_frequency() {
```

Comment:
This test seems to cover a bunch of independent cases. Can we split them up, maybe with a helper function for common setup?